`collect`: don't require a closure

devyn opened the pull request:
# Description

This changes the `collect` command so that it doesn't require a closure. Still allowed, optionally.

Before:

```nushell
open foo.json | insert foo bar | collect { save -f foo.json }
```

After:

```nushell
open foo.json | insert foo bar | collect | save -f foo.json
```

The closure argument isn't really necessary, as collect values are also supported as `PipelineData`.

# User-Facing Changes
- `collect` command changed

# Tests + Formatting
Example changed to reflect.

# After Submitting
- [ ] release notes
- [ ] we may want to deprecate the closure arg?

IanManske commented:
>  we may want to deprecate the closure arg?

I would be in favor of deprecating the closure arg. We would have to figure out another way to implement the `$in` variable before that.

WindSoilder commented:
I don't think it's a breaking change because we convert from a required parameter to optional parameter

WindSoilder commented:
I think we also need to change parser, it uses `collect` command to build `$in` variable.
https://github.com/nushell/nushell/blob/92831d7efcd0586a85966560beb0c538f4a0fc36/crates/nu-parser/src/parser.rs#L6249-L6255

sholderbach commented on crates/nu-cmd-lang/src/core_commands/collect.rs:
```diff
@@ -36,46 +43,52 @@ impl Command for Collect {
+                block,
+                input.into_pipeline_data(),
+            )
+            .map(|x| x.set_metadata(metadata));
```
With this setting the `DataSource::FilePath` again...

sholderbach commented on crates/nu-cmd-lang/src/core_commands/collect.rs:
```diff
@@ -88,7 +101,7 @@ impl Command for Collect {
             Example {
                 description: "Read and write to the same file",
-                example: "open file.txt | collect { save -f file.txt }",
+                example: "open file.txt | collect | save -f file.txt",
```
... I suspect this will fail now (as previously it was running the closure where the metadata was not set)

sholderbach replied:
![grafik](https://github.com/nushell/nushell/assets/15833959/cf8ae2be-1b4d-48fc-87cb-ac45b1f13dff)

IanManske replied:
I think we could maybe remove the same file check in `save` for `PipelineData::Value`. Also, perhaps we should just automatically do a `collect` in `save` if we detect the error condition?

sholderbach replied:
Ignoring it for `PipelineData::Value` sounds like a good idea (until we have lazy data in `Value` coming from a file)

Not sure if I would want to introduce magic into `save` to collect just yet. There may be cases where it alerts you to otherwise bad ideas like round tripping on an otherwise never terminating file/socket etc.?

devyn replied:
Makes sense. I actually assumed that was already the case

sholderbach reviewed (changes requested):
To deliver on the promise for the `open ... | collect | save ...`  round trip we should strip the metadata if it matches `DataSource::FilePath`. The others should probably remain if we don't want to commit to this side effect but I would say that `DataSource::FilePath` is currently an implementation detail for the particular challenge of inplace modifying a file.

devyn commented:
Does this make sense? I changed the `save` command to not check the metadata if it's a `Value`, because there shouldn't be an issue with reading streams in that case

sholderbach commented:
> I changed the `save` command to not check the metadata if it's a `Value`, because there shouldn't be an issue with reading streams in that case

In my view @IanManske's suggestion is an orthogonal improvement. 

I think for this PR on `collect` we can make stronger guarantees for `collect` if you have following lazy operations:

```nushell
open cycle.txt | collect | filter { true } | save cycle.txt
```

devyn commented:
Okay, I think I've fixed it to be correct. File path metadata is removed on `collect`, and it's also set on the input of the closure rather than the output

Added a couple tests to `save` that use `collect` to ensure this works the right way

sholderbach reviewed (approved):
Thanks! Good additional tests
